### DISCUSSION TITLE
Is it possible to force the generation of an OpenApiReference client?

### DISCUSSION THREAD

**Discussion by paulomorgado** [dotnet/aspnetcore#51649] [Q&A]
[_GenerateOpenApiCode](https://github.com/dotnet/aspnetcore/blob/ba3c78b4fbb387d8785dbfb762128a767b9c392c/src/Tools/Extensions.ApiDescription.Client/src/build/Microsoft.Extensions.ApiDescription.Client.targets#L88) is skipped if the output is not outdated.

Is it possible to force it to run?

Any hints, @rainersigwald?

**Comment by danmoseley** [comment]
I'm not aware of any way to tell MSBuild to not skip a target that has no inputs newer than outputs. That is typically done in the authoring by eg deleting the output or by conditionally adding a bogus missing output.

**Reply by rainersigwald** [reply]
Yeah I have occasionally thought of having an engine setting to bypass inputs/outputs and run everything but it doesn't exist today.

I don't understand that target though--if it's basically only calling the MSBuild task, why not drop the target-level incrementality?

**Reply by paulomorgado** [reply]
I added this to my build:

```xml
  <Target Name="__My__GenerateOpenApiCode" BeforeTargets="_GenerateOpenApiCode"
    Condition="$(OpenApiGenerateCodeAtDesignTime) OR ('$(DesignTimeBuild)' != 'true' AND '$(BuildingProject)' == 'true')">
    <Error Condition="!Exists('%(OpenApiReference.FullPath)')" Text="Input OpenAPI file %(Identity) does not exist." />

    <MSBuild Projects="$(MSBuildProjectFullPath)"
        BuildInParallel="$(BuildInParallel)"
        Properties="GeneratorTargetPath=%(OpenApiReference.OutputPath);GeneratorTarget=Generate%(CodeGenerator);GeneratorMetadata=%(SerializedMetadata)"
        Targets="_InnerGenerateOpenApiCode" />
  </Target>
```

and it does the trick. `__My__GenerateOpenApiCode` is always executed (on those conditions) and `_GenerateOpenApiCode` ends up never being executed, because the conditions are the same and, if `__My__GenerateOpenApiCode` ran, outputs are up to date.

**Comment by peteraritchie** [comment]
I ran into this today.  Rebuild didn't help.  I had to do a Clean, then a Build before it would work.  A Clean then a Rebuild, did not. :/

**Reply by paulomorgado** [reply]
That's odd. A Rebuild should be a Clean followed by a Build.

You could collect a binlog to see what's missing.
